[AN-Issue-2002] Extended MultisigTransactionPayload to support GST registration

aregng opened the pull request.

Commit:
[AN-Issue-2002] Updated automation registry to support governance submitted tasks

- Introduced registry configuration and state for system tasks
- Created a new version of the active registry configuration to enclose configurations for both user and governance submitted transactions

TODO:
 - Introduce a new API for system task registration, cancellation
 - Update task processing to handle avoid charging and refunds for system tasks
 - Update already registered task-metadata to mark them as user-tasks during migration
 - Introduce view API to access system task data
 - Introduce API to authorize/deauthorize accounts for system task
   registration

Commit:
Updated automation registry with new API for system task registration

- Introduce a new API for system task registration, cancellation
- Update task processing to handle avoid charging and refunds for system tasks
- Update already registered task-metadata to mark them as user-tasks during migration

Commit:
Added view APIs for system task data

Commit:
Update automation concpets in rust layer according to move layer changes

Commit:
Added unit tests

Commit:
[AN-Issue-2002] Introduced new VM SystemAutomatedTransaction type

- Added new variant to the VM internal transaciton type for system submitted automated transactions
- Extended automted tranaction handler to support system automted transaciton handling
- Extended automated transaction prologue to chack task type as well
- Extended automated transaction epilogue to avoid any fee charges for system automated transactions

Commit:
[AN-Issue-2002] Extended MultisigTransactionPayload to support GST registration

Commit:
Addressed review comments

- Added auxiliary configuration options for easy expansion after release

Commit:
Merge remote-tracking branch 'origin/task/issue-2002-1' into task/issue-2002-2

Commit:
Addressed review comments

Commit:
Merge remote-tracking branch 'origin/task/issue-2002-2' into task/issue-2002-3

sjoshisupra commented on aptos-move/aptos-release-builder/src/components/framework.rs:
```diff
@@ -104,10 +105,10 @@ pub fn generate_upgrade_proposals(
         // If we're generating a single-step proposal on testnet
         if is_testnet && next_execution_hash.is_empty() {
-            release.generate_script_proposal_testnet(account, move_script_path.clone())?;
+            release.generate_script_proposal_testnet(account, move_script_path.clone(), "testnet_script".to_string())?;
```
@aregng it seems all the changes in this file is irrelevant to the PR

aregng replied:
Reverted, will be applied in separate PR.

sjoshisupra commented on aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:
```diff
@@ -556,7 +556,7 @@ impl<'a> AptosTestAdapter<'a> {
             aptos_test_root_address(),
             parameters.sequence_number,
-            aptos_cached_packages::aptos_stdlib::aptos_account_create_account(account_addr),
+            aptos_cached_packages::aptos_stdlib::supra_account_create_account(account_addr),
```
@aregng seems not related to PR

aregng replied:
Reverted, it will be applied in a separate PR.

Commit:
Updated automated transaction creation and added feature gate for system task registration

- Update AutomtionTaskMeta rust type to provide interface for type and priority extraction from aux-data

- Introduce task priority property which is currently initialized with task index value

- Extended RegistrationParams with V2 version which allows to specify task type and priority

- Feature gated task registration via multisig interface

- Feature gated task-type and priority including in aux-data upon registration

- Extended AutomatedTransactionBuilder to take into account task type and priorty

- Updated AutomatedTransactionBuilder result type to be
  AutomatedTransactionDescriptor which encloses type and priority info

- Updated and added unit tests

Commit:
Reverted changes irrelevant to the main commit. They will be applied in a sperate commit

Commit:
Implemented ordering for AutomatedTransactionDescriptor

sjoshisupra commented on aptos-move/aptos-transactional-test-harness/src/aptos_test_harness.rs:
```diff
@@ -556,7 +556,7 @@ impl<'a> AptosTestAdapter<'a> {
             aptos_test_root_address(),
             parameters.sequence_number,
-            aptos_cached_packages::aptos_stdlib::supra_account_create_account(account_addr),
+            aptos_cached_packages::aptos_stdlib::aptos_account_create_account(account_addr),
```
Isn't it `supra_account_create_account`? I believe it refers to Rust interface for `create_account` method of `supra_account` module. @aregng

aregng replied:
Dr. @sjoshisupra , You see this change since the last review, as I have revered the changes that were not related to this PR directly.
It should be `supra.*` and it will be applied in a separate PR along with the other non-relevant changes.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -798,7 +808,7 @@ module supra_framework::automation_registry {
     #[view]
     /// Checks whether there is an active system task in registry with specified input task index.
-    fun has_sender_active_task_with_id_and_type(sender: address, task_index: u64, type: u8): bool acquires AutomationRegistryV2 {
+    public fun has_sender_active_task_with_id_and_type(sender: address, task_index: u64, type: u8): bool acquires AutomationRegistryV2 {
```
I am assuming that another view function where task status is not checked is provided in a different branch/PR to allow one to query owner irrespective of task state.

aregng replied:
Yes, that API is already part of the dev branch, when I have my PRs open merged I will rebase the feature branch to the dev and those changes will be part of the feature branch as well.

sjoshisupra commented on aptos-move/framework/supra-framework/sources/automation_registry.move:
```diff
@@ -3904,6 +3938,78 @@ module supra_framework::automation_registry {
     }
 
+
+    #[test(framework = @supra_framework, user = @0x1cafe)]
+    fun check_task_priority_value(
+        framework: &signer,
+        user: &signer
```
shall we move all tests to a separate `tests/automation_registry_tests.move`?

aregng replied:
Sure, as agreed I will do it in  scope of separate PR.

sjoshisupra commented on aptos-move/e2e-testsuite/src/tests/automation_registration.rs:
```diff
@@ -898,23 +972,97 @@ fn check_automation_registry_actions_when_automation_cycle_disabled() {
+fn find_transaction_error(events: &[ContractEvent]) -> Vec<TransactionExecutionFailed> {
+    events.iter().filter(|e| e.is_v2())
+        .map(|e| bcs::from_bytes::<TransactionExecutionFailed>(e.event_data()))
+        .filter(|d| d.is_ok())
+        .map(|d| d.unwrap())
```
perhaps using `filter_map` might be better here?

Commit:
Merge remote-tracking branch 'origin/feature/detached-automation-cycle' into task/issue-2002-3

Commit:
Addressed review comment and cleaned up after merge

Commit:
Added AutomationTaskMetaData::create function to facilitate testing

sjoshisupra commented on aptos-move/aptos-vm/src/aptos_vm.rs:
```diff
@@ -1116,29 +1138,43 @@ impl AptosVM {
+                            }));
+                        },
+                    }
+                    // TODO: Deduplicate this against execute_multisig_transaction
+                    // A bit tricky since we need to skip success/failure cleanups,
+                    // which is in the middle. Introducing a boolean would make the code
+                    // messier.
```
@aregng file this TODO as issue for improvement/enhancement.

aregng replied:
Dr. @sjoshisupra this comment comes from aptos-source-base (see line 1119 on the left side(source).
Should I still add an enhancement ticket for this?

sjoshisupra replied:
@aregng , oh ok, leave it then.

sjoshisupra commented on types/src/transaction/automation.rs:
```diff
@@ -43,49 +52,150 @@ impl RegistrationParams {
     pub fn automated_function(&self) -> &EntryFunction {
-        let RegistrationParams::V1(v1_self) = self;
-        v1_self.automated_function()
+        match self {
+            RegistrationParams::V1(p) => p.automated_function(),
+            RegistrationParams::V2(p) => p.automated_function(),
+        }
```
Is there a default case for this `match`? If not, I am not sure if `match` is required since in either of the `match` we are performing the same action. In that case removing `match` and just doing `p.automated_function()` may be better.

aregng replied:
There is no a default case. As long as V1 and V2 hold different type of the object although they have similar interface to get the properties we can not apply `|` operator to call the functions.

sjoshisupra commented on types/src/transaction/automation.rs:
```diff
@@ -43,49 +52,150 @@ impl RegistrationParams {
     pub fn expiration_timestamp_secs(&self) -> u64 {
-        let RegistrationParams::V1(v1_self) = self;
-        v1_self.expiration_timestamp_secs
+        match self {
+            RegistrationParams::V1(p) => p.expiration_timestamp_secs,
+            RegistrationParams::V2(p) => p.expiration_timestamp_secs,
+        }
```
same here, if we have same action irrespective of `match` perhaps we directly perform action? If this `match` is needed for identifying appropriate type then can we do
`V1(p) | V2(p)  => p.expiration_timestamp_secs` ?

sjoshisupra commented on types/src/transaction/automation.rs:
```diff
@@ -43,49 +52,150 @@ impl RegistrationParams {
     pub fn function(&self) -> &IdentStr {
-        let RegistrationParams::V1(v1_self) = self;
-        v1_self.function()
+        match self {
+            RegistrationParams::V1(p) => p.function(),
+            RegistrationParams::V2(p) => p.function(),
```
same as above

sjoshisupra commented on types/src/transaction/automation.rs:
```diff
@@ -97,14 +207,27 @@ impl RegistrationParams {
-        let RegistrationParams::V1(v1_self) = self;
-        v1_self.serialized_args_with_sender_and_parent_hash(sender, parent_hash)
+        match self {
+            RegistrationParams::V1(p) => {
+                p.serialized_args_with_sender_and_parent_hash(sender, parent_hash, features)
+            },
+            RegistrationParams::V2(p) => {
+                p.serialized_args_with_sender_and_parent_hash(sender, parent_hash, features)
+            },
```
same as above

sjoshisupra commented on types/src/transaction/automation.rs:
```diff
@@ -97,14 +207,27 @@ impl RegistrationParams {
+    pub fn task_type(&self) -> AutomationTaskType {
+        match self {
+            RegistrationParams::V1(p) => p.task_type(),
+            RegistrationParams::V2(p) => *p.task_type(),
```
is `*` intended in `*p.task_type()`?

aregng replied:
Yes it is require if get function is used, as it returns reference value.

sjoshisupra commented on types/src/transaction/automation.rs:
```diff
@@ -196,19 +299,204 @@ impl RegistrationParamsV1 {
+                .iter()
+                .chain(self.aux_data.iter())
+                .map(|item| MoveValue::vector_u8(item.clone()))
+                .collect()
```
Given that these steps are common, perhaps we only prefix the two values `type_value, priority_value` in the condition and perform the common steps outside of the conditional?

sjoshisupra commented on types/src/transaction/automation.rs:
```diff
@@ -196,19 +299,204 @@ impl RegistrationParamsV1 {
+                .iter()
+                .chain(self.aux_data.iter())
+                .map(|item| MoveValue::vector_u8(item.clone()))
+                .collect()
```
same here, perhaps common steps should be outside of `if`

Commit:
Addressed review comments
